clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -73,6 +77,22 @@ public FetchSessionHandler(LogContext logContext, int node) {
     private LinkedHashMap<TopicPartition, PartitionData> sessionPartitions =
         new LinkedHashMap<>(0);
 
+    /**
+     * All of the topic ids mapped to topic names for topics which exist in the fetch request session.
+     */
+    private Map<String, Uuid> sessionTopicIds = new HashMap<>(0);
+
+    /**
+     * All of the topic names mapped to topic ids for topics which exist in the fetch request session.
+     */
+    private Map<Uuid, String> sessionTopicNames = new HashMap<>(0);
+
+    public Map<Uuid, String> sessionTopicNames() {
+        return sessionTopicNames;
+    }
+
+    private boolean canUseTopicIds = false;
```

Comment:
I think we already do something like this on the broker. We only get to the point of having a session if the broker had an ID for all the topics in the request. I don't think we can calculate on a request basis since we may respond with topics that did not have IDs associated. I may be misunderstanding what you are saying, but I'm very wary of trying to switch between versions 12 and 13 in the same session.

Reply:
> I don't think we can calculate on a request basis since we may respond with topics that did not have IDs associated.

I added another comment in FetchSession. If the session starts with no topicId and a fetch request switches to using topicId, could the server just return an error to force a new session? Will this avoid the need to track canUseTopicIds as a state? Overall, it's probably a bit better to add a bit complexity on the server to simplify the development on the client since we implement the client multiple times in different languages.

Reply:
This is something that we are doing in FetchSession. We close the session if the requests switch between 12 and 13 (or vice versa). Is the idea that we will just send the request based on the topic IDs provided to the builder (if we have an ID for each topic) and let the session code on the server handle it?

Reply:
Yes, if that makes the client code simpler and more consistent. For example, in https://github.com/apache/kafka/pull/9944#discussion_r651933919, the client also chooses to let the server handle the closing of the session.

Reply:
I think the main reason we keep the state in the session for using topic IDs is that some requests may not contain any new/updated partitions and we need to know which version to send. We don't want to close the session and simply send the version that was sent last time. I do think this code is quite confusing as is, so I think I can simply a lot of it.